builder/assignmentEnv.go

```diff
@@ -0,0 +1,315 @@
+}
+
+func (assgnEnv *assignmentEnv) publish() error {
+	if assgnEnv.DockerfileData.Len() <= 0 {
+		return assgnEnv.pullImage()
+	} else if assgnEnv.ImgBuildConfig.publishImage {
+		return assgnEnv.pushImage()
+	}
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) pullImage() error {
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return errors.Wrap(err, "error in creating new docker client")
+	}
+	authConfig := types.AuthConfig{
+		Username: assgnEnv.ImgBuildConfig.authData.Username,
+		Password: assgnEnv.ImgBuildConfig.authData.Password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return errors.Wrap(err, "error in encoding authConfig")
+	}
+
+	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+	resp, err := dockerClient.ImagePull(backgroundContext, assgnEnv.ImgBuildConfig.imageTag, types.ImagePullOptions{
+		RegistryAuth: authStr,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "error in pulling image from hub")
+	}
+	_, err = io.Copy(os.Stdout, resp)
+	if err != nil {
+		return errors.Wrap(err, "error in reading image pull response")
+	}
+	defer func() {
+		err = resp.Close()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+
+	return nil
+
+}
+func (assgnEnv *assignmentEnv) pushImage() error {
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return errors.Wrap(err, "error in creating new docker client")
+	}
+
+	authConfig := types.AuthConfig{
+		Username: assgnEnv.ImgBuildConfig.authData.Username,
+		Password: assgnEnv.ImgBuildConfig.authData.Password,
+	}
+	encodedJSON, err := json.Marshal(authConfig)
+	if err != nil {
+		return errors.Wrap(err, "error in encoding authConfig")
+	}
+
+	authStr := base64.URLEncoding.EncodeToString(encodedJSON)
+
+	imageString := fmt.Sprintf("%s/%s", constants.DockerIO, assgnEnv.ImgBuildConfig.imageTag)
+
+	resp, err := dockerClient.ImagePush(backgroundContext, imageString, types.ImagePushOptions{
+		RegistryAuth: authStr,
+	})
+	if err != nil {
+		return errors.Wrap(err, "error in pushing image to hub")
+	}
+	_, err = io.Copy(os.Stdout, resp)
+	if err != nil {
+		return errors.Wrap(err, "error in reading image push response")
+	}
+	defer func() {
+		err = resp.Close()
+		if err != nil {
+			log.Println(err)
+			return
+		}
+	}()
+	return nil
+}
+
+func (assgnEnv *assignmentEnv) undoWrite() {
+	// Clear the dockerfile data bytes.
+	assgnEnv.DockerfileData.Reset()
+}
+
+func (assgnEnv *assignmentEnv) undoBuild() error {
+	// Delete the created Dockerfile.
+	return os.Remove(assgnEnv.ImgBuildConfig.dockerFilepath)
+}
+
+func (assgnEnv *assignmentEnv) undoPublish() error {
+	// Delete the built image.
+	backgroundContext := context.Background()
+	dockerClient, err := client.NewEnvClient()
+	if err != nil {
+		return err
+	}
+
+	_, err = dockerClient.ImageRemove(backgroundContext, assgnEnv.ImgBuildConfig.imageTag, types.ImageRemoveOptions{
```

Comment:
is this removing the image locally or from docker hub?

Reply:
yes